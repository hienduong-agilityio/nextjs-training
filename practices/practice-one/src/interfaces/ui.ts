export interface IInputGroupProps {
  placeholder?: string;
  value?: string;
  buttonText?: string;
  inputName?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  startIcon?: React.ReactNode;
  customClass?: {
    container?: string;
    inputContainer?: string;
    input?: string;
    button?: string;
  };
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
}
