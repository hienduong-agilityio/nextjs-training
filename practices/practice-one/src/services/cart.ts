// Services
import { apiRequest, getProductById } from '@/services';

// Constants
import { API_URL, HTTP_METHODS } from '@/constants';

// Types
import type { ICart, ICartModifyPayload } from '@/interfaces';

/**
 * Fetch cart by userId
 * @param userId - The ID of the user
 */
export const getCartByUserId = async (userId: number): Promise<ICart> => {
  try {
    const carts = await apiRequest<ICart[]>({
      url: `${API_URL.CART}?userId=${userId}`,
      method: HTTP_METHODS.GET,
    });
    return carts.find((cart: ICart) => cart.userId === userId) || [];
  } catch (error) {
    throw new Error(`Failed to fetch cart for user ${userId}`);
  }
};

/**
 * Add or update a product in the cart
 *
 * @param userId - The ID of the user
 * @param payload - The product details to be added (productId and quantity)
 */
export const addToCart = async (
  userId: number,
  payload: ICartModifyPayload & { maxQuantity: number },
): Promise<{ success: boolean }> => {
  const { maxQuantity } = payload;

  // Fetch cart and product data
  const [cart, productData] = await Promise.all([
    getCartByUserId(userId),
    getProductById(payload.productId),
  ]);

  const discountPercentage = parseFloat(productData.discount ?? '0');

  // Find existing product in the cart
  const existingProduct = cart.products.find(
    (product) => product.id === payload.productId,
  );

  if (existingProduct) {
    const newQuantity = existingProduct.quantity + payload.quantity;

    if (newQuantity > maxQuantity) {
      return { success: false };
    }

    existingProduct.quantity = newQuantity;
    existingProduct.total = existingProduct.price * existingProduct.quantity;
    existingProduct.discountedTotal =
      existingProduct.total * (1 - discountPercentage / 100);
  } else {
    const quantity = Math.min(payload.quantity, maxQuantity);

    cart.products.push({
      id: productData.id,
      title: productData.name,
      price: productData.price,
      quantity,
      total: productData.price * quantity,
      discountPercentage,
      discountedTotal:
        productData.price * quantity * (1 - discountPercentage / 100),
      thumbnail: productData.images[0] || '',
    });
  }

  // Recalculate cart totals
  const { total, discountedTotal, totalQuantity } = cart.products.reduce(
    (acc, product) => {
      acc.total += product.total ?? 0;
      acc.discountedTotal += product.discountedTotal ?? 0;
      acc.totalQuantity += product.quantity;
      return acc;
    },
    { total: 0, discountedTotal: 0, totalQuantity: 0 },
  );

  Object.assign(cart, {
    total,
    discountedTotal,
    totalQuantity,
    totalProducts: cart.products.length,
  });

  // Update cart via API
  await apiRequest<ICart>({
    url: `${API_URL.CART}/${cart.id}`,
    method: HTTP_METHODS.PUT,
    data: cart,
  });

  return { success: true };
};
