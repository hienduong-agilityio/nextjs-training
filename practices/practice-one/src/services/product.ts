// Services
import { apiRequest } from './api';

// Constants
import { API_URL, HTTP_METHODS } from '@/constants';

// Type
import type { IProductProps } from '@/interfaces';

export const getProducts = async (
  category?: string,
): Promise<IProductProps[]> => {
  try {
    return await apiRequest<IProductProps[]>({
      url: category ? `${API_URL.PRODUCT}?${category}` : API_URL.PRODUCT,
      method: HTTP_METHODS.GET,
    });
  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(error.message || 'Failed to fetch products');
    } else {
      throw new Error('An unknown error while fetching products');
    }
  }
};

export const createProduct = async (productData: IProductProps) => {
  try {
    return await apiRequest({
      url: API_URL.PRODUCT,
      method: HTTP_METHODS.POST,
      data: productData,
    });
  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(error.message || 'Failed to create product');
    } else {
      throw new Error('An unknown error while creating the product');
    }
  }
};

export const updateProduct = async (id: string, productData: IProductProps) => {
  try {
    return await apiRequest({
      url: `${API_URL.PRODUCT}/${id}`,
      method: HTTP_METHODS.PUT,
      data: productData,
    });
  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(
        error.message || `Failed to update product with ID ${id}`,
      );
    } else {
      throw new Error('An unknown error while updating the product');
    }
  }
};

export const deleteProduct = async (id: string) => {
  try {
    return await apiRequest({
      url: `${API_URL.PRODUCT}/${id}`,
      method: HTTP_METHODS.DELETE,
    });
  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(
        error.message || `Failed to delete product with ID ${id}`,
      );
    } else {
      throw new Error('An unknown error while deleting the product');
    }
  }
};
