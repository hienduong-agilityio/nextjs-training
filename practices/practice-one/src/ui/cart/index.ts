export * from './CartTable';
export * from './OrderSummary';
export * from './VoucherInput';
export * from './EmptyCart';

export { default as CartItemRow } from './CartItemRow';
