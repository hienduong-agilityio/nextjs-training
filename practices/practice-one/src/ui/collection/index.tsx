export * from './LoadingCategoryGroup';
export * from './LoadingFilterSortBar';
