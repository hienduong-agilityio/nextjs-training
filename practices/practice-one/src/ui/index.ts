export * from './products/index';
export * from './collection/index';
