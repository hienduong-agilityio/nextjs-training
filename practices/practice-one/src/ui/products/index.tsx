export * from './CartAndFavoriteActions';
export * from './ProductImages';
export * from './ProductInfo';
export * from './QuantityControl';
export * from './SocialShare';
export * from './RelatedProduct';
export * from './ProductDetailTabs';
export * from './ProductReview';
export * from './ProductActions';
export * from './ProductLoading';
