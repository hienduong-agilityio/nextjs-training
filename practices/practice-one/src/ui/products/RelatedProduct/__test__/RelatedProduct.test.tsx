import { render, screen } from '@testing-library/react';
import { RelatedProducts } from '@/ui';
import { getProducts, getProductById } from '@/services';

// Mock services
jest.mock('@/services', () => ({
  getProducts: jest.fn(),
  getProductById: jest.fn(),
}));

// Mock component
jest.mock('@/components', () => ({
  ProductCard: jest.fn(({ id, title }) => (
    <div data-testid="product-card" key={id}>
      {title}
    </div>
  )),
}));

describe('RelatedProducts Component', () => {
  const mockProduct = {
    id: '1',
    title: 'Mock Product 1',
    category: 'beauty',
  };

  const mockProducts = [
    { id: '2', title: 'Mock Product 2', category: 'beauty' },
    { id: '3', title: 'Mock Product 3', category: 'beauty' },
    { id: '4', title: 'Mock Product 4', category: 'beauty' },
    { id: '5', title: 'Mock Product 5', category: 'beauty' },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders related products correctly', async () => {
    (getProductById as jest.Mock).mockResolvedValue(mockProduct);
    (getProducts as jest.Mock).mockResolvedValue(mockProducts);

    render(await RelatedProducts({ currentProductId: '1' }));

    expect(getProductById).toHaveBeenCalledWith('1');
    expect(getProducts).toHaveBeenCalledWith({
      page: 1,
      limit: 4,
      filter: { category: 'beauty' },
    });

    const productCards = screen.getAllByTestId('product-card');
    expect(productCards[0]).toHaveTextContent('Mock Product 2');
    expect(productCards[1]).toHaveTextContent('Mock Product 3');
    expect(productCards[2]).toHaveTextContent('Mock Product 4');
  });
});
