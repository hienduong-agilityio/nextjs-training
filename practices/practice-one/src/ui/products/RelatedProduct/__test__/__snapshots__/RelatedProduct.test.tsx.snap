// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedProducts Component matches the snapshot when no related products are found 1`] = `
<div>
  <div
    class="grid w-full grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-6"
  >
    <p>
      No related products found.
    </p>
  </div>
</div>
`;

exports[`RelatedProducts Component matches the snapshot when related products are found 1`] = `
<div>
  <div
    class="grid w-full grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-6"
  >
    <div
      aria-label="Mock Product 2"
    >
      Mock Product 2
    </div>
    <div
      aria-label="Mock Product 3"
    >
      Mock Product 3
    </div>
    <div
      aria-label="Mock Product 4"
    >
      Mock Product 4
    </div>
    <div
      aria-label="Mock Product 5"
    >
      Mock Product 5
    </div>
  </div>
</div>
`;
